pg_wait_sampling.c

```diff
@@ -303,48 +219,45 @@ pgws_shmem_request(void)
 static void
 pgws_shmem_startup(void)
 {
-	bool		found;
-	Size		segsize = pgws_shmem_size();
-	void	   *pgws;
-	shm_toc	   *toc;
+	bool	found;
+	HASHCTL	info;
+
+	if (prev_shmem_startup_hook)
+		prev_shmem_startup_hook();
 
-	pgws = ShmemInitStruct("pg_wait_sampling", segsize, &found);
+	/* Create or attach to the shared memory state */
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
 
+	pgws_proc_queryids = ShmemInitStruct(
+			"pg_wait_sampling queryids",
+			sizeof(pgwsQueryId) * get_max_procs_count(),
+			&found);
+	MemSet(pgws_proc_queryids, 0, sizeof(pgwsQueryId) * get_max_procs_count());
 	if (!found)
 	{
-		toc = shm_toc_create(PG_WAIT_SAMPLING_MAGIC, pgws, segsize);
-
-		pgws_collector_hdr = shm_toc_allocate(toc, sizeof(CollectorShmqHeader));
-		shm_toc_insert(toc, 0, pgws_collector_hdr);
-		pgws_collector_mq = shm_toc_allocate(toc, COLLECTOR_QUEUE_SIZE);
-		shm_toc_insert(toc, 1, pgws_collector_mq);
-		pgws_proc_queryids = shm_toc_allocate(toc,
-									sizeof(uint64) * get_max_procs_count());
-		shm_toc_insert(toc, 2, pgws_proc_queryids);
-		MemSet(pgws_proc_queryids, 0, sizeof(uint64) * get_max_procs_count());
-
-		/* Initialize GUC variables in shared memory */
-		setup_gucs();
-	}
-	else
-	{
-		toc = shm_toc_attach(PG_WAIT_SAMPLING_MAGIC, pgws);
+		/* First time through ... */
+		LWLockPadded *locks = GetNamedLWLockTranche("pg_wait_sampling");
 
-#if PG_VERSION_NUM >= 100000
-		pgws_collector_hdr = shm_toc_lookup(toc, 0, false);
-		pgws_collector_mq = shm_toc_lookup(toc, 1, false);
-		pgws_proc_queryids = shm_toc_lookup(toc, 2, false);
-#else
-		pgws_collector_hdr = shm_toc_lookup(toc, 0);
-		pgws_collector_mq = shm_toc_lookup(toc, 1);
-		pgws_proc_queryids = shm_toc_lookup(toc, 2);
-#endif
+		pgws_profile_lock = &(locks[0]).lock;
+		pgws_history_lock = &(locks[1]).lock;
 	}
 
-	shmem_initialized = true;
+	pgws_history_ring = ShmemInitStruct(
+			"pg_wait_sampling history ring",
+			sizeof(History) + sizeof(HistoryItem) * HistoryBufferSize,
+			&found);
+	pgws_history_ring->index = 0;
 
-	if (prev_shmem_startup_hook)
-		prev_shmem_startup_hook();
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ProfileHashKey);
+	info.entrysize = sizeof(ProfileHashEntry);
+	pgws_profile_hash = ShmemInitHash("pg_wait_sampling hash",
+									  MaxProfileEntries, MaxProfileEntries,
+									  &info, HASH_ELEM | HASH_BLOBS);
```

Comment:
This code should be inside the `if (!found)` branch. Otherwise, this code will be called by every new backend in the EXEC_BACKEND configuration (e.g. on Windows).

Reply:
`ShmemInitHash()` function [assumes](https://github.com/postgres/postgres/blob/393430f57544dfd550135e0191cc91139926b682/src/backend/storage/ipc/shmem.c#L366-L371) idempotent call within itself.